illumos-utils/src/zfs.rs

```diff
@@ -493,6 +567,235 @@ pub struct DatasetEnsureArgs<'a> {
+        let Some(dir_mountpoint) = stdout.trim().lines().next() else {
+            return Err(MountpointError::CheckMountedParse(stdout.to_string()));
+        };
+
+        // If this is a viable mount directory, we'd see something like:
+        //
+        // Root directory:      /
+        // Proposed Mountpoint: /foo
+        //
+        // $ zfs get -Hpo value mountpoint /foo
+        // /
+        //
+        // This means: The mountpoint of "/foo" is the root directory, "/"
+        //
+        // However, if "/foo" was already used as a mountpoint, we'd see:
+        //
+        // $ zfs get -Hpo value mountpoint /foo
+        // /foo
+        //
+        // This means: The mountpoint of "/foo" is "/foo" - it is already
+        // mounted!
+        if dir_mountpoint == mountpoint {
+            return Err(MountpointError::AlreadyMounted);
+        }
+    } else {
+        // If the mountpoint did not exist, create it
+        std::fs::create_dir_all(mountpoint)
+            .map_err(|err| MountpointError::CreateMountpointDirectory(err))?;
+    }
+
+    // "making a directory empty" and "making a directory immutable" cannot
+    // be atomic together. Here's how we cope:
+    //
+    // - If necessary, we make the directory mutable, and try to clear it out.
+    // - We then re-set the directory to immutable.
+    let immutablity = is_directory_immutable(mountpoint)?;
+    if is_directory_empty(mountpoint)?
+        && immutablity.as_immutable_as_filesystem_allows()
+    {
+        return Ok(());
+    }
+
+    if immutablity.can_set_immutable() {
+        make_directory_mutable(mountpoint)?;
+    }
+    ensure_mountpoint_empty(mountpoint)?;
+    if immutablity.can_set_immutable() {
+        make_directory_immutable(mountpoint)?;
+    }
+
+    // This concurrent error case is a bit exceptional: we briefly made the
+    // directory mutable, tried to empty it out, and made it immutable again.
+    // However, while this was happening, someone must have added entries to the
+    // directory.
+    //
+    // This is probably a bug on the side of "whoever is adding these files".
+    if !is_directory_empty(mountpoint)? {
+        return Err(MountpointError::DirectoryNotEmpty);
+    }
+    return Ok(());
+}
+
+fn is_directory_empty(path: &Utf8Path) -> Result<bool, MountpointError> {
+    Ok(path
+        .read_dir_utf8()
+        .map_err(|err| MountpointError::Readdir(err))?
+        .next()
+        .is_none())
+}
+
+fn ensure_mountpoint_empty(path: &Utf8Path) -> Result<(), MountpointError> {
+    if is_directory_empty(path)? {
+        return Ok(());
+    }
+
+    let (Some(parent), Some(file)) = (path.parent(), path.file_name()) else {
+        return Err(MountpointError::BadMountpoint);
+    };
+
+    // The directory is not empty. Let's make a new directory,
+    // with the "old-under-mountpoint-" prefix, and move all data there.
+
+    let prefix = format!("{MOUNTPOINT_TRANSFER_PREFIX}{file}");
```

Comment:
Tiny nit based on Alan's testing - maybe 

```suggestion
    let prefix = format!("{MOUNTPOINT_TRANSFER_PREFIX}{file}-");
```

to put a separator before the random suffix appended by `Utf8TempDir`?

Reply:
Done